jfix-stdlib-ratelimiter/src/main/kotlin/ru/fix/stdlib/ratelimiter/RateLimitedDispatcherProviderImpl.kt

```diff
@@ -0,0 +1,73 @@
+package ru.fix.stdlib.ratelimiter
+
+import ru.fix.aggregating.profiler.Profiler
+import ru.fix.dynamic.property.api.DynamicProperty
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.coroutines.CoroutineContext
+
+class RateLimitedDispatcherProviderImpl(
+    private val rateLimiterFactory: RateLimiterFactory,
+    private val rateLimiterSettings: DynamicProperty<RateLimiterSettings>,
+    private val profiler: Profiler,
+    private val context: CoroutineContext,
+    private val limiterId: String,
+) : RateLimitedDispatcherProvider {
+
+    private var rateLimitedDispatcher: AtomicReference<RateLimitedDispatcherInterface?> = AtomicReference(null)
+    private var isSuspendable = AtomicReference(rateLimiterSettings.get().isSuspendable)
+
+    init {
+        rateLimiterSettings.map { it.isSuspendable }
+            .createSubscription()
+            .setAndCallListener { _, newValue ->
+                if (isSuspendable.get() != newValue) {
+                    updateDispatcher(newValue)
+                }
+            }
+        updateDispatcher(isSuspendable.get())
+    }
+
+    @Synchronized
+    override fun provideDispatcher(): RateLimitedDispatcherInterface = rateLimitedDispatcher.get()!!
+
+    @Synchronized
+    private fun updateDispatcher(newValue: Boolean) {
+        rateLimitedDispatcher.get()?.close()
+        rateLimitedDispatcher.set(
```

Comment:
it is not safe to close dispatcher before setting the new one

current dispatcher might be under pressure from application - such closing will definitely affect request that go through it

I suggest restructure your RateLimitedDispatcherProviderImpl in the following way:

1. lazily create only one instance of each corresponding RateLimitedDispatcher type and store it inside this class
2. make it AutoCloseable and close previusly created initialized dispatchers

For the 1st point you can use kotlin's by `lazy { ... }` (field per dispatcher type) and store only atomic reference to rateLimitedDispatcher . In such way your zk listening update function will:

1. initialize required dispatcher
2. set atomic reference

And provideDispatcher will access atomic reference

Reply:
Done. If I understand the idea correctly